src/Runtime/PhpFpm.php

```diff
@@ -140,31 +142,45 @@ private function eventToFastCgiRequest(array $event): AbstractRequest
         if (! empty($queryString)) {
             $uri .= '?' . $queryString;
         }
-        $request->setRequestUri($uri);
-        $request->setCustomVar('QUERY_STRING', http_build_query($queryParameters));
-
-        $request->setRemoteAddress('127.0.0.1');
-        $request->setRemotePort(80);
-        $request->setServerName('127.0.0.1');
-        $request->setServerPort(80);
-        if (isset($event['requestContext']['protocol'])) {
-            $request->setServerProtocol($event['requestContext']['protocol']);
-        }
+        $queryString = http_build_query($queryParameters);
 
+        $protocol = $event['requestContext']['protocol'] ?? 'HTTP/1.1';
+
+        // Normalize headers
         $headers = $event['headers'] ?? [];
-        if (isset($headers['Host'])) {
-            $request->setServerName($headers['Host']);
+        $headers = array_change_key_case($headers, CASE_LOWER);
+
+        $serverName = $headers['host'] ?? 'localhost';
+
+        $requestHeaders = [
+            'GATEWAY_INTERFACE' => 'FastCGI/1.0',
+            'REQUEST_METHOD' => $event['httpMethod'],
+            'REQUEST_URI' => $uri,
+            'SCRIPT_FILENAME' => $this->handler,
+            'SERVER_SOFTWARE' => 'bref',
+            'REMOTE_ADDR' => '127.0.0.1',
+            'REMOTE_PORT' => '80',
+            'SERVER_ADDR' => '127.0.0.1',
+            'SERVER_NAME' => $serverName,
+            'SERVER_PROTOCOL' => $protocol,
+            'CONTENT_LENGTH' => mb_strlen($requestBody, 'UTF-8'),
```

Comment:
If the content length is expressed as UTF-8 counting, then the content-type must be aware of that assumption. Something like adding `; charset=utf-8` or similar, thoughts?

Reply:
I'm not even sure we should set that every time, see http://php.net/manual/fr/reserved.variables.server.php#110763

@barryvdh I picked that from your implementation, do you know what to do here?

To be honest I have no idea, any suggestion is welcome. Does it make sense to add `; charset=utf-8` to the content-type automatically? The goal would be to stay as close as possible to Apache/Nginx + PHP-FPM.

Reply:
I think you should get them from the HTTP request to be honest :-).